neuralmonkey/encoders/cnn_encoder.py

```diff
@@ -139,17 +255,111 @@ def output(self) -> tf.Tensor:
             dropout_keep_prob=self.dropout_keep_prob,
             train_mode=self.train_mode)
 
-    @tensor
-    def spatial_mask(self) -> tf.Tensor:
-        return tf.ones(tf.shape(self.spatial_states)[:3])
-
     def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
         # if it is from the pickled file, it is list, not numpy tensor,
         # so convert it as as a prevention
         images = np.array(dataset.get_series(self.data_id))
 
         f_dict = {}
         f_dict[self.image_input] = images / 225.0
+        # it is one everywhere where non-zero, i.e. zero columns are masked out
+        f_dict[self.image_mask] = np.sum(
+            np.sign(images), axis=3, keepdims=True)
 
         f_dict[self.train_mode] = train
         return f_dict
+
+
+def max_pooling(
```

Comment:
chápu to dobře že tohle jediný je funkce toho modulu? Jestli jo, dal bych to nahoru a určitě přidal docstring.

Reply:
Je to funkce modulu, ale patři do rodiny s `plain_convolution` a `resnet_block`, ale nemá v sobě self, tak jsem ji vyndal, ať si pylint nestěžuje. Přijde mi lepší, když je blíž těm, ke kterým patří. Navíc mám rád, když otevřu soubor a nahoře je to, co se volá z konfiguráku, ne nějaký helper funkce.

Reply:
ok

Reply:
Já bych to udělal tak, že všechny tři věci budou funkce modulu. `resnet_block` jen dostane batch normalize a train mode, `plain_convolution` jenom train mode. Obojí to potřebujou na batch normalizaci, tak bych jim lambdou dal `batch_normalize_callback`, kde bude train mode vyclosurovaný.

Reply:
stejně to má scaled dot product attention:
https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/attention/scaled_dot_product.py#L95

Reply:
Vůbec se mi to nelíbí, protože to změní čitelnej kód na konstrukci, nad kterou bude muset čtenář kódu dumat, co to vlastně znamená.

Reply:
Já bych zas jako čtenář kódu dumal nad tim, proč je to dvakrát metoda a jednou funkce, respektive mě tahá za oči, jak je to daleko od sebe. Ten `batch_normalize` callback je jen minikonstrukt a každej rozumí tomu, že když se to zavolá, tak se udělá batch normalizace. Plus mi pomůže docstring u těchhle tři funkcí, kde to bude napsaný.